File: packages/plugin-agent/src/server/tools/analysis.ts
import type { AgentToolDefinition, AgentToolHandler, AgentToolContext, AgentToolResult } from '@purplesector/plugin-api';

export const analysisToolDefinitions: AgentToolDefinition[] = [
  {
    name: 'getAnalysisLayout',
    description: 'Get the current analysis panel layout for a given context (e.g. a session or lap).',
    inputSchema: {
      type: 'object',
      properties: {
        context: { type: 'string', description: 'Layout context string, e.g. "global", "session:<id>", "lap:<id>"' },
      },
      required: ['context'],
    },
    category: 'analysis',
    mutating: false,
  },
  {
    name: 'listSavedAnalysisLayouts',
    description: 'List all saved analysis layouts for the current user.',
    inputSchema: { type: 'object', properties: {} },
    category: 'analysis',
    mutating: false,
  },
  {
    name: 'saveAnalysisLayout',
    description: 'Save the current analysis layout with a name.',
    inputSchema: {
      type: 'object',
      properties: {
        name: { type: 'string', description: 'Layout name' },
        description: { type: 'string' },
        layout: { type: 'string', description: 'JSON string of the AnalysisLayoutJSON' },
        context: { type: 'string', description: 'Layout context (default: "global")' },
        isDefault: { type: 'boolean', description: 'Whether this should be the default layout for the context' },
      },
      required: ['name', 'layout'],
    },
    category: 'analysis',
    mutating: true,
  },
];

export function createAnalysisToolHandlers(prisma: any): Record<string, AgentToolHandler> {
  return {
    async getAnalysisLayout(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const layout = await prisma.savedAnalysisLayout.findFirst({
        where: { userId: ctx.userId, context: args.context as string, isDefault: true },
      });
      if (!layout) {
        return { success: true, data: null, message: 'No saved layout found for this context. The app uses a default layout.' };
      }
      return { success: true, data: { ...layout, layout: JSON.parse(layout.layout) } };
    },

    async listSavedAnalysisLayouts(_args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const layouts = await prisma.savedAnalysisLayout.findMany({
        where: { userId: ctx.userId },
        select: { id: true, name: true, description: true, context: true, isDefault: true, createdAt: true },
        orderBy: { createdAt: 'desc' },
      });
      return { success: true, data: layouts, message: `Found ${layouts.length} saved layouts.` };
    },

    async saveAnalysisLayout(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const context = (args.context as string) || 'global';
      const isDefault = args.isDefault === true;

      if (isDefault) {
        await prisma.savedAnalysisLayout.updateMany({
          where: { userId: ctx.userId, context, isDefault: true },
          data: { isDefault: false },
        });
      }

      const layout = await prisma.savedAnalysisLayout.create({
        data: {
          userId: ctx.userId,
          name: args.name as string,
          description: (args.description as string) || null,
          layout: args.layout as string,
          context,
          isDefault,
        },
      });
      return { success: true, data: layout, message: `Saved layout "${layout.name}".` };
    },
  };
}

File: apps/web/apps/agent/tools/analysis.py
"""Agent tools for saved analysis layouts."""

from __future__ import annotations

import json

from apps.analysis.models import SavedAnalysisLayout

from . import ToolDefinition, ToolContext, ToolResult, register_tools

definitions = [
    ToolDefinition(
        name="getAnalysisLayout",
        description="Get the current analysis panel layout for a given context (e.g. a session or lap).",
        input_schema={
            "type": "object",
            "properties": {
                "context": {"type": "string", "description": 'Layout context string, e.g. "global", "session:<id>", "lap:<id>"'},
            },
            "required": ["context"],
        },
        category="analysis",
    ),
    ToolDefinition(
        name="listSavedAnalysisLayouts",
        description="List all saved analysis layouts for the current user.",
        input_schema={"type": "object", "properties": {}},
        category="analysis",
    ),
    ToolDefinition(
        name="saveAnalysisLayout",
        description="Save the current analysis layout with a name.",
        input_schema={
            "type": "object",
            "properties": {
                "name": {"type": "string", "description": "Layout name"},
                "description": {"type": "string"},
                "layout": {"type": "string", "description": "JSON string of the AnalysisLayoutJSON"},
                "context": {"type": "string", "description": 'Layout context (default: "global")'},
                "isDefault": {"type": "boolean", "description": "Whether this should be the default layout for the context"},
            },
            "required": ["name", "layout"],
        },
        category="analysis",
        mutating=True,
    ),
]


async def get_analysis_layout(args: dict, ctx: ToolContext) -> ToolResult:
    layout = (
        SavedAnalysisLayout.objects.filter(user_id=ctx.user_id, context=args["context"], is_default=True)
        .first()
    )
    if not layout:
        return ToolResult(success=True, data=None, message="No saved layout found for this context. The app uses a default layout.")
    try:
        parsed = json.loads(layout.layout)
    except (json.JSONDecodeError, TypeError):
        parsed = layout.layout
    return ToolResult(
        success=True,
        data={"id": str(layout.id), "name": layout.name, "description": layout.description, "layout": parsed, "context": layout.context},
    )


async def list_saved_analysis_layouts(args: dict, ctx: ToolContext) -> ToolResult:
    layouts = SavedAnalysisLayout.objects.filter(user_id=ctx.user_id).order_by("-created_at")
    data = [
        {
            "id": str(l.id),
            "name": l.name,
            "description": l.description,
            "context": l.context,
            "isDefault": l.is_default,
            "createdAt": l.created_at.isoformat(),
        }
        for l in layouts
    ]
    return ToolResult(success=True, data=data, message=f"Found {len(data)} saved layouts.")


async def save_analysis_layout(args: dict, ctx: ToolContext) -> ToolResult:
    context = args.get("context") or "global"
    is_default = args.get("isDefault", False)

    if is_default:
        SavedAnalysisLayout.objects.filter(user_id=ctx.user_id, context=context, is_default=True).update(is_default=False)

    layout = SavedAnalysisLayout.objects.create(
        user_id=ctx.user_id,
        name=args["name"],
        description=args.get("description") or None,
        layout=args["layout"],
        context=context,
        is_default=is_default,
    )
    return ToolResult(success=True, data={"id": str(layout.id), "name": layout.name}, message=f'Saved layout "{layout.name}".')


handlers = {
    "getAnalysisLayout": get_analysis_layout,
    "listSavedAnalysisLayouts": list_saved_analysis_layouts,
    "saveAnalysisLayout": save_analysis_layout,
}

register_tools(definitions, handlers)
